DHCP Daemon

dougbtv opened the pull request:
Adds rendering, bindata & unit tests for the containernetworking/plugins DHCP CNI plugin running as a daemon.

Currently has no dependency on the macvlan functionality, however, it anticipates it with a commented-out `TODO` block.

danwinship commented on bindata/network/dhcp/dhcp-daemon.yaml:
```diff
@@ -0,0 +1,63 @@
+        - "-hostprefix"
+        - "/host"
+        securityContext:
+          privileged: true
```
why does this need to be privileged?

danwinship commented on bindata/network/dhcp/dhcp-daemon.yaml:
```diff
@@ -0,0 +1,63 @@
+        command: ["/bin/sh"]
+        args: ["-c", "rm -f /var/run/cni/dhcp.sock"]
+        securityContext:
+          privileged: true
```
even assuming the `dhcp-daemon` container needs to be privileged, why does the `initContainer` need to be privileged? (also, needs a better `name`)

danwinship commented on pkg/network/dhcp_daemon_test.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// TestRenderNoDHCP tests a rendering with the DHCP daemonset.
+func TestRenderWithDHCP(t *testing.T) {
```
comment should say `TestRenderWithDHCP`

danwinship commented on bindata/network/dhcp/dhcp-daemon.yaml:
```diff
@@ -0,0 +1,63 @@
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+  name: dhcp-daemon
```
could this just be added to the existing multus DaemonSet as an optional additional container? (using `{{if ...}}` to optionally include/exclude part of the template based on a variable set by the renderer, as in `bindata/network/openshift-sdn/004-multitenant.yaml`)

danwinship commented on pkg/network/render.go:
```diff
@@ -318,3 +326,76 @@ func RenderMultus(conf *operv1.NetworkSpec, manifestDir string) ([]*uns.Unstruct
+func RenderDHCP(conf *operv1.NetworkSpec, manifestDir string) ([]*uns.Unstructured, error) {
+
+	// This isn't useful with Multinetwork.
+	if *conf.DisableMultiNetwork {
```
isn't useful *without* MultiNetwork

danwinship commented on pkg/network/render.go:
```diff
@@ -318,3 +326,76 @@ func RenderMultus(conf *operv1.NetworkSpec, manifestDir string) ([]*uns.Unstruct
+	if conf.AdditionalNetworks != nil {
+
+		for _, addnet := range conf.AdditionalNetworks {
+
```
(all these blank lines at the starts of blocks are non-idiomatic...)

danwinship commented on pkg/network/render.go:
```diff
@@ -318,3 +326,76 @@ func RenderMultus(conf *operv1.NetworkSpec, manifestDir string) ([]*uns.Unstruct
+
+			// Cycle through the IPAM keys, and determine if the type is dhcp
+			if rawConfig["ipam"] != nil {
+				ipam := rawConfig["ipam"].(map[string]interface{})
```
This will panic if the CNI is invalid (`"ipam": "yes"`). Need to do a checked typecast.

Likewise below. Alternatively, demarshal it into a real type and make `json.Unmarshal` do the checking for you:

    type CNINetConfIPAMOnly struct {
    	IPAM struct {
    		Type string `json:"type"`
    	} `json:"ipam"`
    }

dougbtv commented:
@danwinship -- thanks a bunch for the thorough review, appreciate it.

I pushed a new commit, I believe I covered what you've suggested, including:

* Removed privileged security context (it's not necessary)
* Updated name of init container
* Updated comments per suggestions.
* Moved DHCP yaml to Multus dir and have it handled by Multus renderer (nice suggestion)
* Removed blank lines at beginning of functions.
* Checked for errors during type conversion, added test with bogus IPAM definition.

Thanks again.

danwinship commented on bindata/network/multus/003-dhcp-daemon.yaml:
```diff
@@ -0,0 +1,60 @@
+{{if .RenderDHCP}}
```
So I actually meant to put this inside `multus.yaml`; rather than deploying an additional DaemonSet, with an additional Pod on each node, just stick the containers here into the existing multus daemonset, conditionally on `RenderDHCP` being set.

(This could be done as an optimization after the initial PR merges maybe though.)

dougbtv replied:
Hey Dan -- what's the best way to test for the presence of a certain named container in the unit tests? I saw there's a custom matcher that does it for specific `kind:` but, was having trouble testing for the presence of a specific container within a daemonset.

danwinship commented on bindata/network/multus/003-dhcp-daemon.yaml:
```diff
@@ -0,0 +1,60 @@
+            path: /var/run/cni
+        - name: procpath
+          hostPath:
+            path: /proc
```
hm... so I was wondering why we need this, and the answer is, so that the daemon can nsenter into pod netnamespaces and fiddle with their network interfaces, except that in that case, probably it did actually need to be privileged?

dougbtv replied:
Ahhh yes, I had started to wonder why I had the privileged clause in my original sample daemonset from PoCs

danwinship commented on pkg/network/dhcp_daemon.go:
```diff
@@ -0,0 +1,57 @@
+
+				for key, value := range ipam {
+					if key == "type" {
+						if value.(string) == "dhcp" {
```
The `for key, value` doesn't help here; you're still doing an unsafe cast. (eg,  it would crash on `{"ipam": { "type": 5 }}`

dougbtv replied:
Great catch, missed that one, thanks.

dougbtv commented:
@danwinship -- got another suite of changes in.

* Fixed the missed check on the type casting. 
* Returned privileged to the dhcp-daemon container
* Also tweaked the UseDHCP() method so that it doesn't error but just warn (otherwise it'd prevent Multus from starting at all just due to some bad data, for example)
* Did not address moving from daemonset for DHCP to containers within Multus -- that should be fine, but, I was having trouble validating it in my tests.

Thanks again for the reviews.

danwinship commented:
> Did not address moving from daemonset for DHCP to containers within Multus -- that should be fine, but, I was having trouble validating it in my tests.

Yeah, probably needs a new helper function. We can do this later.

danwinship commented:
/lgtm

danwinship commented:
/retest

dougbtv commented:
/retest
